proposals/001_resource_kubernetes_patch.md

```diff
@@ -0,0 +1,165 @@
+- Modifying the default namespace/service account on a cluster without importing / recreating
+
+Previously we viewed this functionality as more of a "carry out this operation with `kubectl` before running terraform" thing, but given the mounting list of use cases and user appetite I think this fits our workflows based approach and we should try and accommodate.
+
+
+# Proposed Configuration
+
+A generic `kubernetes_patch` resource should allow you specify:
+
+- The kind of resource being patched
+- The resource's metadata (i.e name and namespace) 
+- The type of patch operation 
+- The patch itself 
+
+In the kubernetes provider, it would look like this. 
+
+```hcl
+resource "kubernetes_patch" {
+  kind = "namespace"
+
+  metadata {
+    name = "default"
+  }
+ 
+  patch = jsonencode({
+    metadata = {
+      labels = {
+        test = "test"
+      }
+    }
+  })
+ 
+  type = "merge"
+}
+```
+
+In the kubernetes-alpha provider:
+
+```hcl
+resource "kubernetes_patch" {
+  provider = kubernetes-alpha
+
+  kind = "namespace"
+
+  metadata {
+    name = "default"
+  }
+ 
+  patch = {
+    metadata = {
+      labels = {
+        test = "test"
+      }
+    }
+  }
+ 
+  type = "merge"
+}
+```
+
+## Possible Alternatives
+
+A possible alternative to the generic patch resource above is to implement use-case specific resources for managing things like annotations, tolerations, and so on. 
+
+For example:
+
+```hcl
+resource "kubernetes_annotations" {
+  kind = "pod"
+
+  metadata {
+    name = "pod-name"
+  }  
+
+  annotations = {
+    "test-annotation" = "value"
+  }
+}
+```
+
+The big downside of this is increased maintenance burden – we will have to implement a new resource for every use case that merits a resource being patched.  
+
+
+## Discussion
+
+
+1. **Should we implement this in the kubernetes provider, or in the kubernetes-alpha provider?**
+
+Arguments _for_ kubernetes provider:
+- Stability – users of the main provider won’t need to add kubernetes-alpha which still in flux. 
+- Simple patch doesn’t have as many exotic challenges as custom resources.
+- Easier implementation than in the alpha provider because it uses the main Terraform SDK. 
+
+Arguments _against_ kubernetes provider:
+- The patch has to be stored as a text string and used with `jsondecode`/`jsonencode`.
+- We will have to do JSON munging to do any validation on the patch.
+
+
+
+Arguments _for_ kubernetes-alpha provider:
+- Cleaner syntax, no need for `jsonencode()`.
+- Adding a highly wanted feature that incentivizes people to adopt the kubernetes-alpha provider.
+- Code already exists in kubernetes-alpha to do things like validate against the OpenAPI schema at plan time.
+
+Arguments _against_ kubernetes-alpha provider:
+- Scope of the alpha provider becomes less tightly circumscribed around custom resources
+- More conservative users will have to wait for the alpha provider to hit GA 
+
+
+2. **Should we support the JSON patch type of just strategic and merge?**
```

Comment:
JSON patch seems more deterministic. Also, seems to be the only can describe differential operations, allowing to distinguish between intentions like: "add / remove THIS element to THAT list" and "replace THIS list with THAT list" (same for maps).